### DISCUSSION TITLE
How to get x,y coordinates of atoms from diagram using javascript RDKit MinimalLib?

### DISCUSSION THREAD

**Discussion by heberleh** [rdkit/rdkit#4844] [Q&A] [answered]
Is there a way to get x,y coordinates from atoms of a diagram using the javascript implementation?

I could see only the functions `mol.get_new_coords()` and `mol.get_molblock()` which seem to do the same thing and the output contains potentially the x,y coordinates.

Are those coordinates referring to the diagram coordinates? (it feels like they don't)
If so, how can we convert them to fit the canvas dimensions?

I'm considering using RDKit  in a project that is currently using the smilesDrawer package. In the smilesDrawer implementation, x,y coordinates before and after creating and attaching the <canvas> to the DOM are different, they are scaled and added some offset.

What would be the case for rdkit?

In the end, I need a <canvas> element with the drawing and the atoms' coordinates, to create layers of new <canvas> over the original diagram.

Many thanks and happy new year!

**Comment by ptosco** [comment]
@heberleh the coordinates returned by `mol.get_new_coords()` and `mol.get_molblock()` are in angstrom.
These are converted to diagram coordinates once they are painted onto an HTML5 `canvas` using `mol.draw_to_canvas()` or to a SVG string using `get_svg()`. Diagram coordinates can be extracted from the SVG string; it will be helpful to get rid of the atom labels to simplify the drawing; you can do so by passing appropriate `drawOpts`, e.g.
```
const drawOpts = {noAtomLabels: true, width: 200, height: 200};
const svg = mol.get_svg_with_highlights(JSON.stringify(drawOpts))
const parser = new DOMParser();
const svgXml = parser.parseFromString(svg, 'text/xml');
const pathNodes = svgXml.querySelectorAll('path');
```
I hope this helps to get you started.

**Comment by DavidACosgrove** [comment]
It's worth noting, however, that the coordinates for the drawing without
the labels may be different from those with labels, if there is a label on
the extremes of the picture, because the drawing scale will different in
the two cases.  When establishing the scale to convert the angstrom
coordinates to drawing coordinates, it adjusts for atom labels if
necessary. The best way to achieve what you want would be to call the
getDrawCoords(atIdx) method of the underlying MolDraw2DJS object that
produces the drawing, but I don't know if that is exposed to the JS or just
used in the background.

On Mon, Jan 3, 2022 at 3:19 PM Paolo Tosco ***@***.***> wrote:

> @heberleh <https://github.com/heberleh> the coordinates returned by
> mol.get_new_coords() and mol.get_molblock() are in angstrom.
> These are converted to diagram coordinates once they are painted onto an
> HTML5 canvas using mol.draw_to_canvas() or to a SVG string using get_svg().
> Diagram coordinates can be extracted from the SVG string; it will be
> helpful to get rid of the atom labels to simplify the drawing; you can do
> so by passing appropriate drawOpts, e.g.
>
> const drawOpts = {noAtomLabels: true, width: 200, height: 200};
> const svg = mol.get_svg_with_highlights(JSON.stringify(drawOpts))
> const parser = new DOMParser();
> const svgXml = parser.parseFromString(svg, 'text/xml');
> const pathNodes = svgXml.querySelectorAll('path');
>
> I hope this helps to get you started.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/rdkit/rdkit/discussions/4844#discussioncomment-1900638>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACGF2FQIELLIZ7PF42CUB2TUUG5B5ANCNFSM5LFHSHKA>
> .
> Triage notifications on the go with GitHub Mobile for iOS
> <https://apps.apple.com/app/apple-store/id1477376905?ct=notification-email&mt=8&pt=524675>
> or Android
> <https://play.google.com/store/apps/details?id=com.github.android&referrer=utm_campaign%3Dnotification-email%26utm_medium%3Demail%26utm_source%3Dgithub>.
>
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>


-- 
David Cosgrove
Freelance computational chemistry and chemoinformatics developer
http://cozchemix.co.uk

**Comment by heberleh** [comment, marked as answer]
Because currently there is no function to find the coordinates, use the instructions by @ptosco from here https://github.com/rdkit/rdkit/discussions/4844#discussioncomment-1900638

then process the SVG as follows:

The SVG is divided in Vertex paths (atoms, drawings of letters) and Edges paths (lines from atom to atom).

Find a position of atoms from Vertex paths:

1. Identify the vertex nodes by filtering by classList size === 1
2. Get coordinates by parsing the path.d points
3. Calculated Convex Hull (https://www.npmjs.com/package/convexhull-js) of all points
4. Get average x and y

Find a position of other atoms based on Edges paths:

1. A bond can be represented by 1, 2, or more paths... for example, a double bond can be represented by 2 or 4 paths
2. Unfortunately, the 4 (or 2, 6, ...) paths have the same "ids" (classes) so you can't know directly what is what, you need to process.
3. Find if they are parallel or if they connect with each other, to understand if they are "parts" or double/triple bonds... Trick: parts usually have the same size and connect with each other. If it is a ring, the parallel edges... one will be shorter than the other, get the ratio and filter by <0.99. Another thing is that the longer edge seems to be always the first in the SVG. Double bond formed by 4 paths will also follow the sequence: (part1, part2) bond1, (part1, part2) bond1. 


All together:

1. Get atoms indexes from classList.item(0) => atom-<index>
2. For each index, filter the paths that contain this index in any of the classList items (id, source or target)
3. Give preference for Vertex paths; if found -> return x,y
4. Then give preference for Single bonds; if found -> return x,y 
5. Then try to hack remaining paths if no single bond was found or vertex-path... which means you need to join find what are the source/target from each edge... and which path is longer (in case of rings). Again, usually the first joined path is the longest one from the rings.
